CondTools/Geometry/plugins/PPSGeometryBuilder.cc

```diff
@@ -0,0 +1,264 @@
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/ESWatcher.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/Exception.h"
+
+#include "FWCore/ServiceRegistry/interface/Service.h"
+#include "CondCore/DBOutputService/interface/PoolDBOutputService.h"
+
+#include "DetectorDescription/Core/interface/DDCompactView.h"
+#include "DetectorDescription/Core/interface/DDFilteredView.h"
+
+#include "DataFormats/CTPPSDetId/interface/TotemRPDetId.h"
+#include "DataFormats/CTPPSDetId/interface/TotemTimingDetId.h"
+#include "DataFormats/CTPPSDetId/interface/CTPPSPixelDetId.h"
+#include "DataFormats/CTPPSDetId/interface/CTPPSDiamondDetId.h"
+#include "Geometry/VeryForwardGeometryBuilder/interface/CTPPSDDDNames.h"
+
+#include "Geometry/Records/interface/IdealGeometryRecord.h"
+#include "Geometry/Records/interface/VeryForwardIdealGeometryRecord.h"
+
+#include "CondFormats/GeometryObjects/interface/PDetGeomDesc.h"
+
+#include <regex>
+
+//----------------------------------------------------------------------------------------------------
+
+class PPSGeometryBuilder : public edm::one::EDAnalyzer<> {
+public:
+  explicit PPSGeometryBuilder(const edm::ParameterSet&);
+  void analyze(const edm::Event&, const edm::EventSetup&) override;
+
+private:
+  void buildPDetGeomDesc(DDFilteredView*, PDetGeomDesc*);
+  uint32_t getGeographicalID(DDFilteredView*);
+
+  std::string compactViewTag_;
+  edm::ESWatcher<IdealGeometryRecord> watcherIdealGeometry_;
+  edm::Service<cond::service::PoolDBOutputService> dbservice_;
+};
+
+//----------------------------------------------------------------------------------------------------
+
+PPSGeometryBuilder::PPSGeometryBuilder(const edm::ParameterSet& iConfig)
+    : compactViewTag_(iConfig.getUntrackedParameter<std::string>("compactViewTag", "XMLIdealGeometryESSource_CTPPS")) {}
+
+//----------------------------------------------------------------------------------------------------
+
+void PPSGeometryBuilder::analyze(const edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  edm::ESHandle<DDCompactView> cpv;
+
+  // Get DDCompactView from IdealGeometryRecord
+  if (watcherIdealGeometry_.check(iSetup)) {
+    iSetup.get<IdealGeometryRecord>().get(compactViewTag_.c_str(), cpv);
+  }
+
+  // Create DDFilteredView and apply the filter
+  DDPassAllFilter filter;
+  DDFilteredView fv((*cpv), filter);
+
+  // Persistent geometry data
+  PDetGeomDesc* pdet = new PDetGeomDesc;
+  // Build geometry
+  buildPDetGeomDesc(&fv, pdet);
+
+  // Save geometry in the database
+  if (pdet->container_.empty()) {
+    throw cms::Exception("PPSGeometryBuilder") << "PDetGeomDesc is empty, no geometry to save in the database.";
+  } else {
+    if (dbservice_.isAvailable()) {
+      dbservice_->writeOne(pdet, dbservice_->beginOfTime(), "VeryForwardIdealGeometryRecord");
+    } else {
+      throw cms::Exception("PoolDBService required.");
+    }
+  }
+
+  return;
```

Comment:
This `return` statement is not needed.

Reply:
Hi Carl, as it is geometry-approved now, should I remove the return statement and commit it again? Or leave as it is and correct when I migrate it do DD4hep?

Reply:
If any of the other reviewers ask for changes, you can make this fix along with other changes. If nothing else changes in this PR, you can fix this little statement in a later PR.